### ISSUE TITLE
When pytest xdist is used, pytest logs are not output to the console and log files. Is this block compatible

### ISSUE THREAD

**Issue by chineseluo** [pytest-dev/pytest-xdist#574]
When pytest xdist is used, pytest logs are not output to the console and log files. Is this block compatible

**Comment by tetienne** [comment]
Any workaround about this?

**Comment by nicoddemus** [comment]
Unfortunately this is a known limitation: due to how execnet works (the backend for process spawning used by xdist), stdout and stderr cannot be sent to the master node.

**Comment by andreabisello** [comment]
i never found a solution for this, so when i need to obtain test results as fast as possible, i use xdist, when i need logs, i don't use xdist, running one test at time.

**Comment by seanv507** [comment]
I feel like this is a standard problem when doing parallel logging. It feels like there is a standard solution using eg https://pytest-xdist.readthedocs.io/en/latest/how-to.html#creating-one-log-file-for-each-worker.

Is anyone more experienced able to provide a working solution?

**Comment by RonnyPfannschmidt** [comment]
Currently there is no safe way to live log parallel runs as log output from multiple tests would painfully intersperse and confuse

Nobody so far came up with a usable solution for that on plain pytest

**Comment by seanv507** [comment]
@RonnyPfannschmidt agreed, but many people would be happy without *live* logs. eg reporting log files consecutively, after the jobs have finished.  This is what I imagined would be a "standard solution", and I was imagining it would help many people if there was some standard example of eg detecting if xdist is running, saving logfiles to a file and then printing them/saving them as artifacts of the build (in eg Jenkins)

I assume many people do not know this package intimately, but just look for a quick way to make their test runs faster and then hit this problem, and as @andreabisello said just give up on logging.

**Comment by RonnyPfannschmidt** [comment]
Quick inline/at test error /output reporting may be a nice feature for pytest core

So far im not aware of any proposal in that direction

Also if there ever was a textual based ux it could manage the data

**Comment by andreabisello** [comment]
as any worker logs his logs,
with this hook (to be adapted) we join all the workers logs in a single file.

```

def pytest_sessionfinish(session):
    lista = []
    nome_file_log = str(nome_modulo_per_logfile) + '.log'

    for file in os.listdir('logs/workers/'):
        if fnmatch.fnmatch(file, 'tmp-pytest-*'):
            with open('logs/workers/' + str(file)) as f:
                f = f.readlines()
            for line in f:
                lista.append(line)

    with open('logs/' + nome_file_log, 'w+') as f:
        for items in lista:
            f.write('%s' % items)
    for file in os.listdir('logs/workers/'):
        if fnmatch.fnmatch(file, 'tmp-pytest-*') and file != 'tmp-pytest-master.log':
            os.remove('logs/workers/' + str(file))

```

**Comment by RonnyPfannschmidt** [comment]
@anfreshman what you mention sound like the normal output capture of pytest  - in particular as it seems like in your case xdist is not involved at all

**Comment by anfreshman** [comment]
> @anfreshman what you mention sound like the normal output capture of pytest - in particular as it seems like in your case xdist is not involved at all

I apologize for providing too vague a description, in fact, I use pytest-xdist and the phenomenon is closely related to it
Phenomenon: Use the pytest command to directly start the pytest project. At the same time, after adding addopts = -n auto in the pytest.ini file, run the pytest case directly on the console. The console will no longer output the log information printed through logging inside the case.
Environment configuration: The running machine is MacBook Air, and the chip is Apple M2. The pytest version used is 7.1.1 and the pytest-xdist version is 2.5.0
I also submitted a minimal reproduction environment. If necessary, you can refer to this project. There is only one case and you can run the reproduction directly.     [https://github.com/anfreshman/pytestLogExample](url)

**Comment by anfreshman** [comment]
> @anfreshman你提到的听起来像是 pytest 的正常输出捕获 - 特别是在你的情况下 xdist 似乎根本没有参与

After debugging for a while, I think I figured out the reasons for some of the related phenomena:
1. Phenomenon: When using -n auto and setting parameterization for case, the related output streams are not printed to the console
Reason: In fact, they (including stdout and stderr) are captured by the capture component of the pytest framework
But I noticed that when "--capture=no" is set, stderr will be output normally, but stdout still cannot be output normally. There is a similar phenomenon described in the known limitations of the pytest-xidst document https://pytest-xdist.readthedocs.io/en/stable/known-limitations.html, but the document expects that stderr and stdout will not be output normally, which is still somewhat different from the experimental results.
In addition, there are some less intuitive phenomena, such as:
When pytest-xidst is introduced without the "-n auto" command, stdout and stderr can also be output normally in the console when they are captured
I will continue debugging to try to understand what happened, but if you already understand the reason and are willing to share it, I will be very grateful

**Comment by sguerrero11** [comment]
I started using xdist recently and it seems you can print everything you want in the console but for some reason logging into a file doesn't work the same way. I guess pytest doesn't know how to make a distinction between workers and thus it will just save the tests results into the log file to avoid some sort of conflict...this is weird as the console logging works fine, it doesn't make any sense that you cannot save the same into a file.

However, using just 1 worker results in the same, so that's a contradiction. Logging into a file doesn't log all output when using xdist, that's the conclusion and I'm not sure there's a way to bypass this.
